MU plugin for the Server-Timing API

mukeshpanchal27 opened the pull request:
The PR adds a MU plugin for Server-Timing API and a GH action step that add mu plugin in docker container.

Closes #3

mukeshpanchal27 committed:
Add MU plugin for the Server-Timing API

mukeshpanchal27 committed:
Merge branch 'feature/automated-performance-testing-mvp' into add/server-timing-api-mu-plugin

felixarntz commented on tests/performance/plugins/server-timing/mu-plugin.php:
```diff
@@ -0,0 +1,47 @@
+					$server_timing_values = array();
+				}
+
+				$server_timing_values['template'] = microtime( true ) - $timestart;
```
This is slightly off, as right now it measures the total time. It needs to only measure the time starting from the above `template_redirect` hook.

You can define a variable like `$template_start` in that hook callback that stores the current `microtime( true )` and then use that in here instead of the `$timestart` global, which is only needed for the `before-template` metric.

mukeshpanchal27 replied:
Thanks for the review and pointing it out. I have address this in [b2581a7](https://github.com/10up/wordpress-develop/pull/7/commits/b2581a70662d9917e23094359c0e2ce47dcd467d)

felixarntz commented on tests/performance/plugins/server-timing/mu-plugin.php:
```diff
@@ -0,0 +1,47 @@
+		}
+
+		$server_timing_values['before-template'] = microtime( true ) - $timestart;
+
```
Nit-pick, but I'd advise to start the output buffer at the end of this, right here.

felixarntz commented on tests/performance/plugins/server-timing/mu-plugin.php:
```diff
@@ -0,0 +1,47 @@
+			function() {
+
+				global $server_timing_values, $timestart;
+
```
See above nit-pick, maybe best to end the output buffer right away here.

felixarntz commented on tests/performance/plugins/server-timing/mu-plugin.php:
```diff
@@ -0,0 +1,47 @@
+
+				global $server_timing_values, $timestart;
+
+				if ( ! is_array( $server_timing_values ) ) {
+					$server_timing_values = array();
+				}
```
This shouldn't be needed strictly speaking, since the variable is also ensured to be an array above. If it wasn't an array here, it also would be missing the `before-template` key which is unconditionally checked further below.

mukeshpanchal27 replied:
Yes. Make sense.

felixarntz commented on tests/performance/plugins/server-timing/mu-plugin.php:
```diff
@@ -0,0 +1,47 @@
+			-9999
+		);
+	},
+	-9999
```
This hook should rather happen as late as possible, since we only want to start the output buffer as close as possible to the template actually being rendered. For example, other hook callbacks potentially added should run before this one.

So I would suggest to change this to `9999`.

joemcgill replied:
Perhaps `PHP_INT_MAX` instead of `9999`?

mukeshpanchal27 replied:
Thanks for the feedback, I have updated priority for `template_redirect` to `PHP_INT_MAX`

felixarntz reviewed (commented):
@mukeshpanchal27 Looks like this includes the code from #6 and will need to be refreshed once that PR is merged.

But I already took a look at the new code in this PR and left some feedback. Looks almost good to go!

joemcgill commented:
> Looks like this includes the code from https://github.com/10up/wordpress-develop/pull/6

To make reviews easier, you could set the target branch to the [setup-initial-performance-action](https://github.com/10up/wordpress-develop/tree/setup-initial-performance-action) branch and mark this a blocked until after #6 is merged. That way no refactor will be needed and this PR should automatically be updated to point to the feature branch once that one gets merged.

mukeshpanchal27 committed:
Address review feedback

joemcgill reviewed (approved):
The updates to the mu-plugin look good to me.

felixarntz commented on tests/performance/plugins/server-timing/mu-plugin.php:
```diff
@@ -0,0 +1,45 @@
+	'template_redirect',
+	function() {
+
+		global $server_timing_values, $timestart, $template_start;
```
Nit-pick: Better not to make `$template_start` a global variable, as it's only needed in here and the closure below. You can pass it to the other closure via `use()`.

Maybe we should even keep `$server_timing_values` as just a local variable for now. At least for now, the way this code is written there is no real benefit in having it as a global. That would change if we want to add server timing values from elsewhere, but at least now we don't, so maybe best to avoid the global.

mukeshpanchal27 replied:
Thanks @felixarntz Good catch.

felixarntz commented on tests/performance/plugins/server-timing/mu-plugin.php:
```diff
@@ -0,0 +1,45 @@
+			'shutdown',
+			function() {
+
+				global $server_timing_values, $timestart, $template_start;
```
See above, here you could access the local variables instead of using globals, e.g. like:
```suggestion
			function() use ( $server_timing_values, $template_start ) {
```

mukeshpanchal27 replied:
Thanks for the suggestion.

felixarntz reviewed (commented):
@mukeshpanchal27 One small point of feedback on avoiding globals where not necessary. That in itself is not necessarily a blocker, but now that #6 was merged, we need to also refresh this branch and resolve the merge conflict, so maybe you can address my new feedback then as well.

Related is @joemcgill's comment from https://github.com/10up/wordpress-develop/pull/7#issuecomment-1412754388, that way we can avoid such conflicts in the future.

mukeshpanchal27 committed:
Address review feedback

mukeshpanchal27 committed:
Merge branch 'feature/automated-performance-testing-mvp' into add/server-timing-api-mu-plugin

mukeshpanchal27 commented:
Thanks for the feedback. @joemcgill I have addressed your feedback, which you have asked for [here](https://github.com/10up/wordpress-develop/pull/6#pullrequestreview-1281165914).

joemcgill commented:
Thanks @mukeshpanchal27. The updates look good to me.

felixarntz commented on .github/workflows/performance.yml:
```diff
@@ -67,12 +62,19 @@ jobs:
+      - name: Install MU plugin
+        run: |
+          mkdir ./src/wp-content/mu-plugins
+          cp ./tests/performance/plugins/server-timing/mu-plugin.php ./src/wp-content/mu-plugins/server-timing.php
```
Nit-pick: It would make more sense to me to have this file in `./tests/performance/mu-plugins/server-timing.php`, as that would indicate better where it should go. Seems unnecessary to use a different structure in the source directory for this MU plugin.

mukeshpanchal27 replied:
Thanks @felixarntz for the review. https://github.com/10up/wordpress-develop/pull/7/commits/85a5c96fe2bbccc806c4b8eb06863a26c4285617 updates the MU plugin path.

felixarntz reviewed (approved):
@mukeshpanchal27 Thank you for making the iterations, LGTM. Tiny feedback below, but not blocking!

mukeshpanchal27 committed:
Update MU plugin path
